TRD raw data reconstruction, compressed and not.

bazinski opened the pull request:
- read cru on flp
- compress
- read raw and compressed data on epn.
- data is self consistant with the mc 2 raw generation.
- zero suppressed digits (mc is only done with these)
- real raw data needs the byte swap option described in the docs.
- add event record which stores a trd event.
Documentation added to the readme as well.

bazinski committed:
cru data to tracklets and digits and compress

fixed the swapping options for mc vs raw

bazinski commented:
I just realised that zero surpressed digits are not handled. Fix coming.

bazinski commented:
@martenole you can take a look now, triggerrecords fixed. There are redudant stuff in some of the classes, ignore those for now, they will come out eventually.

martenole commented on DataFormats/Detectors/TRD/include/DataFormatsTRD/CompressedDigit.h:
```diff
@@ -0,0 +1,96 @@
+#include <numeric>
+#include "Rtypes.h" // for ClassDef
+#include "DataFormatsTRD/Constants.h"
+#include "gsl/span"
```
not sure if all these includes are needed? vector, unordered_map, numeric and span I don't see, but maybe I miss them

bazinski replied:
ja there is a pile of historical includes, was planning on refactoring during next week.
fixed.

martenole commented on DataFormats/Detectors/TRD/src/CompressedDigit.cxx:
```diff
@@ -0,0 +1,54 @@
+  setROB(rob);
+  setMCM(mcm);
+  setChannel(channel);
+  //setADC(adc);
```
why not setting the adcs here?

bazinski replied:
was actually not using that part constructor but mapping the digits in via some pointer arithmetic.

martenole commented on Detectors/TRD/reconstruction/README.md:
````diff
@@ -0,0 +1,51 @@
+```
+o2-sim -n 200 -g pythia8 --skipModules ZDC
+o2-sim-digitizer-workflow -b
+o2-trd-trap-sim -b 
````
Actually not needed anymore, since by default TRAP is run within digitizer workflow :)

bazinski replied:
i remember that discussion but thought the pr was reverted ... cool

martenole commented on Detectors/TRD/reconstruction/src/DataReader.cxx:
```diff
@@ -0,0 +1,86 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   raw2tracklet.cxx
```
typo

bazinski replied:
more than likely more of those. thanks for spotting.

martenole commented on Detectors/TRD/reconstruction/src/CruCompressor.cxx:
```diff
@@ -0,0 +1,101 @@
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+/// @file   raw2tracklet.cxx
```
typo

bazinski replied:
guess how I did the license thing at the top ;-) will fix.

martenole reviewed (dismissed):
@bazinski thanks, I only have very minor comments from looking at the code. But I also did not go into the implementation details..

bazinski commented:
@martenole  regarding the implementation details, its going to be easier just to test it. I am trying to formulate my tests to be reasonably full coverage, getting there.

davidrohr commented:
Hi @bazinski : I have tried to run this, but I see 2 issues:
- The `o2-trd-datareader` doesn't start, since the `trd-datareader-inputspec` option does not exist. I am a bit confused, since there is an option `trd-datareader-config` which defaults to `TRD/RAWDATA` but then it is not used. Perhaps this is just a typo, and the config option should be the inputspec?
- When I manually set the inputspec to `TRD/RAWDATA`, I can run it, but I get the following warning and then a segfault:
```
[110163:trd-datareader]: setting buffer position to 0x7f20493b5280
[110163:trd-datareader]: [22:46:01][INFO]  Setting buffer size to : 43232
[110163:trd-datareader]: [22:46:01][INFO] And away we go, run method of Translator
[110163:trd-datareader]: [22:46:01][INFO] rewinding crurawreader incoming data buffer
[110163:trd-datareader]: [22:46:01][INFO] EP:0 CRU:0x0000 Link:15  FEEID:0x0000 SrcID:TRD[4] Packet:1   MemSize:64    OffsNext:64     prio.:0 HS:64 HV:6
[110163:trd-datareader]: [22:46:01][INFO] Orbit:0         BC:0    Stop:1 Page:1     Trg:00000000000000000000100010000011 Par:0     DetFld:0x0000
[110163:trd-datareader]: [22:46:01][INFO] Tracklet Parser parse of data sitting at :0x7f20c1550048 starting at pos 0x7f20c1550088
[110163:trd-datareader]: [22:46:01][INFO]  index is :  fffffff0 word is :0x7f20c1550088  start is : 0x7f20c1550088 endis : 0x7f20c158a988
[110163:trd-datareader]: [22:46:01][INFO] Before byteswapping -16 word is : 0x7f20c1550088 next word is : 0 and raw nextword is :0
[110163:trd-datareader]: [22:46:01][INFO] Before byteswapping -16 word is : 61710 next word is : 0 and raw nextword is :0
[110163:trd-datareader]: [22:46:01][INFO] Before byteswapping -16 word is : 0x7f20c1550088 next word is : 0 and raw nextword is :0
[110163:trd-datareader]: [22:46:01][INFO] tracklet parsing 0x 0x7f20c1550088 at pos : 0
[110163:trd-datareader]: [22:46:01][WARN] mcmtrackletcount is not in [0:2] something very wrong parsing the TrackletMCMData fields with data of : 0x61710
```

The stack trace for the segfault is:
```
#0  o2::trd::getQFromRaw (header=0x0, data=0x7f5a3bc89088, pidindex=pidindex@entry=0, trackletindex=trackletindex@entry=0) at /home/qon/alice/sw/SOURCES/O2/dev/0/DataFormats/Detectors/TRD/src/RawData.cxx:117
#1  0x00007f5a49a15fc8 in o2::trd::TrackletsParser::Parse (this=this@entry=0x7f5a3c0891a0) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/TrackletsParser.cxx:153
#2  0x00007f5a49a180a6 in o2::trd::TrackletsParser::Parse (dataverbose=false, headerverbose=false, verbose=true, cleardigits=false, detector=1, end=0x7f5a3bcc3988, start=0x7f5a3bc89088, data=0x7f5a3bc89048, this=0x7f5a3c0891a0) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/include/TRDReconstruction/TrackletsParser.h:45
#3  o2::trd::CruRawReader::processHalfCRU (this=0x7f5a3bc89028) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:211
#4  0x00007f5a49a19ff3 in o2::trd::CruRawReader::processHBFs (this=0x7f5a3bc89028, datasizealreadyread=<optimized out>, verbose=<optimized out>) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:127
#5  0x00007f5a49a1b613 in o2::trd::CruRawReader::run (this=this@entry=0x7f5a3bc89028) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:298
#6  0x00007f5a49a20975 in o2::trd::DataReaderTask::run (this=0x7f5a3bc89020, pc=...) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/DataReaderTask.cxx:81
```

Steps to reproduce:
```
o2-sim -n 10 -g pythia8
o2-sim-digitizer-workflow
o2-trd-trap-sim --disable-mc
o2-trd-trap2raw -o raw/TRD --fileper cru
o2-raw-file-reader-workflow --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader
```

Could you have a look?

bazinski commented:
That is indeed weird, looking ...   I see verbose output is still hard coded to default on.

davidrohr commented:
@bazinski : Any news?
Also, this must be rebased after The movement of the TRD headers into DataFormats.

bazinski commented:
@davidrohr something came up, not putting it here.
As for the dataformat header files its only digits.h header in question as far as i know, but ja will rebase with some pending updates.

bazinski commented:
@davidrohr ok this might take a while to fix, there is a fundamental problem with the structure of the parsing data. It works for the current raw data I have, but its wrong ....  the rawwriter produces "real" raw data. Long story, will fix the full system test dependent parts first.

bazinski committed:
fix link lengths from mc

bazinski committed:
zero suppressed mc2raw

fix parsing zero suppressed digits

increase sanity check on trd cru link lengths

bazinski commented:
o2-sim -n 10 -g pythia8
o2-sim-digitizer-workflow
o2-trd-trap-sim --disable-mc
o2-trd-trap2raw -o raw/TRD --fileper halfcru
o2-raw-file-reader-workflow --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader --trd-datareader-disablebyteswapdata 


-n 100 I get the tracklet(s) without digits, correctly detected as a fatal error. I suspect something wrong with my simulation, I have 6 other simulations that work fine.

number of digits and tracklets match up and in verbose modes the outputs can matched off. Will show on Monday.
--trd-datareader-verbose --trd-datareader-dataverbose --trd-datareader-headerverbose.
grep output for TTT for tracklets and DDDD for digits, slice and dice and sort and files are the same. tests cases to come.

bazinski commented:
for the 100 event dataset that is causing the problems : 

...
Digit : 48033  hcid=1069 calculated hcid=1069 det=534 mcm=2 rob=5 ch=4  | Tracklet : 5217 hcid=1069 det=534 mcm=3 rob=3·
Digit : 48034  hcid=1069 calculated hcid=1069 det=534 mcm=2 rob=5 ch=5  | Tracklet : 5218  hcid=1069 det=534 mcm=3 rob=3·
Digit : 48035  hcid=1069 calculated hcid=1069 det=534 mcm=3 rob=5 ch=20 | Tracklet : 5219  hcid=1069 det=534 mcm=2 rob=5·
Digit : 48036  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=2   | Tracklet : 5220  hcid=22 det=11 mcm=2 rob=2·
Digit : 48037  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=3   | Tracklet : 5221  hcid=22 det=11 mcm=2 rob=2·
Digit : 48038  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=4   | Tracklet : 5222  hcid=25 det=12 mcm=15 rob=3·
Digit : 48039  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=5   | Tracklet : 5223  hcid=25 det=12 mcm=3 rob=5·
Digit : 48040  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=6   | Tracklet : 5224  hcid=25 det=12 mcm=3 rob=5
Digit : 48041  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=7   | Tracklet : 5225  hcid=25 det=12 mcm=3 rob=5·
Digit : 48042  hcid=683 calculated hcid=683 det=341 mcm=13 rob=5 ch=8   | Tracklet : 5226  hcid=25 det=12 mcm=7 rob=5·
Digit : 48043  hcid=683 calculated hcid=683 det=341 mcm=14 rob=5 ch=20  | Tracklet : 5227  hcid=59 det=29 mcm=9 rob=1·
Digit : 48044  hcid=10 calculated hcid=10 det=5 mcm=5 rob=4 ch=2        | Tracklet : 5228  hcid=59 det=29 mcm=9 rob=1·
Digit : 48045  hcid=10 calculated hcid=10 det=5 mcm=5 rob=4 ch=3        | Tracklet : 5229  hcid=59 det=29 mcm=9 rob=1·
Digit : 48046  hcid=10 calculated hcid=10 det=5 mcm=5 rob=4 ch=4        | Tracklet : 5230 hcid=59 det=29 mcm=10 rob=1·
Digit : 48047  hcid=10 calculated hcid=10 det=5 mcm=5 rob=4 ch=5        | Tracklet : 5231 hcid=76 det=38 mcm=5 rob=0·  
.....

left is digit vector, right is tracklet vector and they the are seperately walked through and matched off, by hcid/mcm/rob.
digit 48036 has hcid 683 and should not be there. 48044 with hcid having looped is clearly the next interaction record and should be walked through trying to matchoff to tracklet 5220. 48036 should either be further up above hcid 1069 or further down beyond hcid 10 ...
Why this only happens for the 100 even example is also a mystery.
This is of course happening at the beginning part of 
o2-trd-trap2raw -o raw/TRD --fileper halfcru


*update, those hcid 683 are infact their own triggerrecord, the sorting is correct.  A triggerrecord of 8 digits and zero tracklets.
Code does not take account of a triggerrecord with no tracklets!
The code assumes the triggerrecord in the tracklets file are matched 1 for 1.
So its skips over the zero tracklet triggerrecord and then they go out of sink .... more to fix

bazinski committed:
case of zero tracklets in triggerrecord fixed

reverse boolean logic for byteswapping trd payload data

bazinski commented:
fixed

davidrohr commented:
> fixed

@bazinski : Could you clarify what exactly is fixed? I mean, shall I retest this for the full system test now?

bazinski commented:
@davidrohr  :-D sorry ja that was not clear, lots of things are fixed over the duration of this pr, the "fixed" comment above is in reference to the crashon 100 events from this morning mentioned in a meeting at 1 this afternoon (the stuff directly above).

Yes the commands you typed above should work now. It was working fine yesterday until the bigger simulation I did last night to test against. Ironic that a zero tracklet trigger only comes with the bigger sim.

you need to add --trd-datareader-disablebyteswapdata to the datareader, I have only tested "--fileper halfcru" I assume that makes no difference to the fullsystem test? I will go through test the various other options later this evening and do unit tests for it all.

bazinski commented:
blank raw files fixed.

bazinski committed:
reverse boolean logic for byteswapping trd payload data

martenole commented:
Hi Sean,
I just tried again your commands from above and it seems to work fine with
`o2-trd-trap2raw -o raw/TRD --fileper halfcru`
`o2-raw-file-reader-workflow -b --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader -b --trd-datareader-enablebyteswapdata`.
I just compared the number of digits and tracklets found in the raw stream with what was simulated and it looks consistent.
If I remove the flag `trd-datareader-enablebyteswapdata` I get a segfault at `/home/oschmidt/alice/sw/SOURCES/O2/rawdataread/0/DataFormats/Detectors/TRD/src/RawData.cxx:117`. Is it possible to catch the wrong order somehow?
The build errors seem to be insufficient memory on the build machines :/

davidrohr commented:
@bazinski : The raw converter is not reading the configKeyValues or the config file. Could you apply the following patch:
```
diff --git a/Detectors/TRD/simulation/src/trap2raw.cxx b/Detectors/TRD/simulation/src/trap2raw.cxx
index 99db4d91ee8..abaf98ecb21 100644
--- a/Detectors/TRD/simulation/src/trap2raw.cxx
+++ b/Detectors/TRD/simulation/src/trap2raw.cxx
@@ -74,6 +74,7 @@ int main(int argc, char** argv)
     add_option("no-empty-hbf,e", bpo::value<bool>()->default_value(false)->implicit_value(true), "do not create empty HBF pages (except for HBF starting TF)");
     add_option("rdh-version,r", bpo::value<uint32_t>()->default_value(6), "rdh version in use default");
     add_option("configKeyValues", bpo::value<std::string>()->default_value(""), "comma-separated configKeyValues");
+    add_option("hbfutils-config,u", bpo::value<std::string>()->default_value(std::string(o2::base::NameConf::DIGITIZATIONCONFIGFILE)), "config file for HBFUtils (or none)");
     add_option("reject-digits", bpo::value<bool>()->default_value(false), "reject the digits in the raw stream");
     add_option("verbose,w", bpo::value<bool>()->default_value(false), "verbose");
 
@@ -95,7 +96,11 @@ int main(int argc, char** argv)
     std::cerr << e.what() << ", application will now exit" << std::endl;
     exit(2);
   }
-  //  o2::conf::ConfigurableParam::updateFromString(vm["configKeyValues"].as<std::string>());
+  std::string confDig = vm["hbfutils-config"].as<std::string>();
+  if (!confDig.empty() && confDig != "none") {
+    o2::conf::ConfigurableParam::updateFromFile(confDig, "HBFUtils");
+  }
+  o2::conf::ConfigurableParam::updateFromString(vm["configKeyValues"].as<std::string>());
 
   std::cout << "yay it ran" << std::endl;
   trap2raw(vm["input-file-digits"].as<std::string>(), vm["input-file-tracklets"].as<std::string>(), vm["output-dir"].as<std::string>(), vm["verbosity"].as<int>(),
```
Otherwise, we cannot apply settings such as `HBFUtils.nHBFPerTF=128`.

davidrohr commented:
Hi @bazinski 
I have tried it now with 5 Pb-Pb events, and with a larger set of pp events. I am getting a segfault with stacktrace:
```
#0  0x00007fa26d86e64e in o2::trd::TrackletsParser::Parse (this=this@entry=0x7fa25f6151b0) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/TrackletsParser.cxx:243
#1  0x00007fa26d87451a in o2::trd::TrackletsParser::Parse (dataverbose=<optimized out>, headerverbose=<optimized out>, verbose=<optimized out>, disablebyteswap=<optimized out>, cleardigits=false, detector=1, end=0x7fa25f215188, start=0x7fa25f2150c8, data=0x7fa25f215048, this=0x7fa25f6151b0) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/include/TRDReconstruction/TrackletsParser.h:51
#2  o2::trd::CruRawReader::processHalfCRU (this=0x7fa25f215028, cruhbfstartoffset=16) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:257
#3  0x00007fa26d87b2e8 in o2::trd::CruRawReader::processHBFs (this=0x7fa25f215028, datasizealreadyread=<optimized out>, verbose=<optimized out>) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:153
#4  0x00007fa26d87c4e5 in o2::trd::CruRawReader::run (this=this@entry=0x7fa25f215028) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/CruRawReader.cxx:378
#5  0x00007fa26d881dad in o2::trd::DataReaderTask::run (this=0x7fa25f215020, pc=...) at /home/qon/alice/sw/SOURCES/O2/dev/0/Detectors/TRD/reconstruction/src/DataReaderTask.cxx:92
```

This is by just running
```
o2-sim -n 10 -g pythia8
o2-sim-digitizer-workflow
o2-trd-trap-sim --disable-mc
o2-trd-trap2raw -o raw/TRD --fileper halfcru
o2-raw-file-reader-workflow --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader --trd-datareader-enablebyteswapdata
```
I also tried without the `--trd-datareader-enablebyteswapdata` option, and also that segfaults.
I have uploaded all my files here: https://qon.jwdt.org/nmls/tmp/trd-raw-testcase.tar.bz2

bazinski commented:
Hi @davidrohr 

I know there is a problem at the moment with the tracklet half chamber header pointer that if the first file read contains a first event without a tracklet it segfaults. Busy fixing. Only one of my examples has this.

Thanks for the patch, will put in.

bazinski committed:
fix half chamber header info arriving late

bazinski commented:
@davidrohr my segfault fixed. I think yours was due to you byteswapping.

o2-sim -n 10 -g pythia8
o2-sim-digitizer-workflow
o2-trd-trap-sim --disable-mc
o2-trd-trap2raw -o raw/TRD --fileper halfcru
o2-raw-file-reader-workflow --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader 

Default is not to byteswap anymore, real raw data will now need the switch though.

davidrohr commented:
@bazinski : With your new patch, my data seems to work, thx.
I am now doing some more tests. Is this PR then be ready t be merged? O2/o2 CI is failling generally due to memory, but if the fullCI passes, I would like to merge this asap, so we can progress with TRD in the FST.

bazinski commented:
@davidrohr we are both desperate to get this off our plates ;-)

Yes please merge at your convenience. I suppose @martenole @jolopezl or @tdietel should approve ?

Unit tests will come later and compressor verification/tests.

martenole commented on Detectors/TRD/workflow/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@
 # submit itself to any jurisdiction.
 
-#add_compile_options(-O0 -pg -fPIC)
+add_compile_options(-O0 -pg -fPIC)
```
Should this be kept in?

martenole commented on Detectors/TRD/reconstruction/CMakeLists.txt:
```diff
@@ -8,9 +8,17 @@
 #granted to it by virtue of its status as an Intergovernmental Organization or
 #submit itself to any jurisdiction.
 
+add_compile_options(-O0 -g -fPIC)
```
Also here?

bazinski replied:
ah oops .......  shot for catching

martenole reviewed (dismissed):
Just not sure if the compile flags should be added to those TRD modules

davidrohr commented:
Thx for catching, obviously they should not

davidrohr commented:
@bazinski @martenole : I have tried to use the `o2-trd-datareader` in combination with the tracklet transformer on a 128 orbit TF from the full system test:
```
o2-raw-file-reader-workflow --input-conf raw/TRD/TRDraw.cfg | o2-trd-datareader |o2-trd-tracklet-transformer --shm-segment-size 32000000000 --root-in 0 --root-out 0
```
That fails since the tracklet transformer doesn't find an output spec for `TRD/TRACKLETS/0`. I have applied the following patch to the datareader, which I guess should also go into this PR.
```
diff --git a/Detectors/TRD/reconstruction/src/DataReader.cxx b/Detectors/TRD/reconstruction/src/DataReader.cxx
index 7551c78a81d..5b521b2cfc0 100644
--- a/Detectors/TRD/reconstruction/src/DataReader.cxx
+++ b/Detectors/TRD/reconstruction/src/DataReader.cxx
@@ -63,10 +63,10 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   auto dataverbose = cfgc.options().get<bool>("trd-datareader-dataverbose");
 
   std::vector<OutputSpec> outputs;
-  outputs.emplace_back(OutputSpec(ConcreteDataTypeMatcher{"TRD", "TRACKLETS"}));
-  outputs.emplace_back(OutputSpec(ConcreteDataTypeMatcher{"TRD", "DIGITS"}));
-  outputs.emplace_back(OutputSpec(ConcreteDataTypeMatcher{"TRD", "TRIGGERRECORD"}));
-  //outputs.emplace_back(OutputSpec(ConcreteDataTypeMatcher{"TRD", "FLPSTAT"}));
+  outputs.emplace_back("TRD", "TRACKLETS", 0, Lifetime::Timeframe);
+  outputs.emplace_back("TRD", "DIGITS", 0, Lifetime::Timeframe);
+  outputs.emplace_back("TRD", "TRIGGERRECORD", 0, Lifetime::Timeframe);
+  //outputs.emplace_back("TRD", "FLPSTAT", 0, Lifetime::Timeframe);
   LOG(info) << "input spec is:" << inputspec;
   LOG(info) << "enablebyteswap :" << byteswap;
   AlgorithmSpec algoSpec;
```

Afterwards, the workflow runs, but in the log I see:
```
[31135:trd-datareader]: [13:09:39][INFO] TRD Translator Task run
[31135:trd-datareader]: [13:09:55][INFO] Processing time for Data reading  16366ms
[31135:trd-datareader]: [13:09:55][INFO] Digits found : 28651701
[31135:trd-datareader]: [13:09:55][INFO] Digits returned : 28651701
[31135:trd-datareader]: [13:09:55][INFO] Tracklets found : 3116702
[31135:trd-datareader]: [13:09:55][INFO] Tracklets returned : 3116702
[31136:TRDTRACKLETTRANSFORMER]: [13:10:00][INFO] running tracklet transformer
[31136:TRDTRACKLETTRANSFORMER]: [13:10:00][INFO] 2483 tracklets found!
```
So the tracklet transformer processes way less tracklets than the datareader.

Running the tracklet transformer on the tracklets without raw conversion via
```
o2-trd-tracklet-transformer --shm-segment-size 32000000000
```
I see
```
[32158:TRDTRACKLETTRANSFORMER]: [13:12:25][INFO] running tracklet transformer
[32158:TRDTRACKLETTRANSFORMER]: [13:12:25][INFO] 3301316 tracklets found!
```
So the same order of magnitude, but the number of tracklets is not the same.

Finally, if I run the o2 raw file check on the TRD raw files I see errors:
```
[O2sim/latest] ~/alitest/tmp-fst $> o2-raw-file-check raw/TRD/TRDraw.cfg 
[INFO] RawDataHeader v6 is assumed
[INFO] apply  check for Mismatch between flagged and calculated new TF start
[INFO] apply  check for No SOX found on 1st page
[INFO] ignore check for TF does not start by new superpage
[INFO] apply  check for Wrong HBF orbit increment
[INFO] apply  check for Number of TFs is less than expected
[INFO] apply  check for Number of HBFs per TF not as expected
[INFO] apply  check for Data does not start with TF/HBF
[INFO] apply  check for New HBF starts w/o closing old one
[INFO] ignore check for RDH.stop set of 1st HBF page
[INFO] apply  check for Wrong RDH.pageCnt increment
[INFO] apply  check for Wrong RDH.packetCounter increment
[INFO] perform check for /Wrong RDH.packetCounter increment/
[INFO] perform check for /Wrong RDH.pageCnt increment/
[INFO] perform check for /New HBF starts w/o closing old one/
[INFO] perform check for /Data does not start with TF/HBF/
[INFO] perform check for /Number of HBFs per TF not as expected/
[INFO] perform check for /Number of TFs is less than expected/
[INFO] perform check for /Wrong HBF orbit increment/
[INFO] perform check for /No SOX found on 1st page/
[INFO] perform check for /Mismatch between flagged and calculated new TF start/
[ERROR] Data does not start with TF/HBF
[ERROR]  ^^^Problem(s) was encountered at offset 0 of file 0
[ERROR] Unexpected RDH version 35 from
```

I will retry this now with shorter time frames, but it seems there are still some bugs in there.

davidrohr commented:
OK, quick update, using a smaller time frame, I see the same error in the raw file check, and the number of tracklets received by the transformer is still not correct:
```
[3009:trd-datareader]: [13:21:14][INFO] Tracklets found : 55298
[3009:trd-datareader]: [13:21:14][INFO] Tracklets returned : 55298
[3010:TRDTRACKLETTRANSFORMER]: [13:21:14][INFO] running tracklet transformer
[3010:TRDTRACKLETTRANSFORMER]: [13:21:14][INFO] 788 tracklets found!
```
I think the problem that in the large time frame the number of tracklets from the root file and from the raw file do not match is due to the raw file containing only 128 orbits, and the ROOT file more. So please ignore this.

bazinski commented:
i thought the time frame is pre my code hence that one line that was commented out and in your patch was put back in. All the timing information is in the trigger records. Maybe i am missing something. i suspect the error you seeing about the tracklet count is actually problems with the trigger records. investigating ....

bazinski commented:
@davidrohr is o2-raw-file-checker supposed to parse a raw config file and find the files?

if do 
o2-raw-file-check raw/TRD/*.raw 
it seems happy

davidrohr commented:
> @davidrohr is o2-raw-file-checker supposed to parse a raw config file and find the files?
> 
> if do
> o2-raw-file-check raw/TRD/*.raw
> it seems happy

You have to give it the cfg file, like with the raw reader:
```
o2-raw-file-check raw/TRD/TRDraw.cfg 
```

bazinski commented:
@davidrohr am i missing something does o2-raw-file-check not require the -c ?
the following works fine for me: 
o2-raw-file-check -c raw/TRD/TRDraw.cfg 

because the error message says the first file does not start with an rdh but the first file is fine on hex 
raw/TRD/trd_cru_35_h.raw

davidrohr commented:
> @davidrohr am i missing something does o2-raw-file-check not require the -c ?
> the following works fine for me:
> o2-raw-file-check -c raw/TRD/TRDraw.cfg
> 
> because the error message says the first file does not start with an rdh but the first file is fine on hex
> raw/TRD/trd_cru_35_h.raw

Oh, sure you are right, my fault :(
So I think the only problem is that the trackletconverter gets too few tracklets. Sorry...

bazinski commented:
@martenole the tracklettransformer is only receivng the very first message. Am I sending it wrong somehow or does tracklettransformer only ever accept a single message ? Is this @jbarrella 's code?

davidrohr commented:
@bazinski @martenole : I am not sure, but I believe at least the TRD tracker expects the tracklets as a continuous array for the time frame. So if the raw decoder sends them not as a single message, they would need to be rearranged afterwards I guess.

davidrohr commented:
@bazinski : This is a real error: `/mnt/mesos/sandbox/sandbox/o2-fullci/sw/SOURCES/O2/5912/0/Detectors/TRD/base/src/FeeParam.cxx:402:19: error: statement should be inside braces [readability-braces-around-statements]`

bazinski commented:
@davidrohr shot for the heads up, i have gotten so used to ignoring the red crosses.

bazinski committed:
trd event record, overhaul events

davidrohr commented:
@bazinski : I think you forgot to add a file to git

bazinski commented:
@davidrohr shot actually forgot to remove the include and file, its from a previous iteration of this. Just sorting out the bunch crossing, i have 2 things called bunchcrossing and only in very rare cases are they the same :-( rdh vs cru header.

bazinski committed:
change some fatal messages to warn

bazinski commented:
@davidrohr 1 remaining problem with digits not being attached to messages but that wont affect anything you want to do.

davidrohr commented:
ok, thx, giving it a try...

davidrohr commented:
Hi @bazinski , I checked the new version, and it was ridiculously slow, > 10 minutes per time frame, with 96% of the time spent in memcpy and 2% in malloc according to perf. This was due to some extensive copying due to improper usage of range-based for loops. Not sure if there are more cases, but the attached patch improves the timing from > 10 minutes to 1.8 seconds on my system. [trd.patch.gz](https://github.com/AliceO2Group/AliceO2/files/6636576/trd.patch.gz)

The number of tracklets is correct now, I will make some more tests with the full system test now.

Can you include my patch, check other range based loops, and fix the conflict due to @martenole 's PR which was merged.
If afterwards you let me know that you are ready, and if I don't see further problems in the FST, we can finally merge this.

davidrohr commented:
Hi @bazinski : Please also apply this patch to the PR:
```
diff --git a/Detectors/TRD/reconstruction/src/DataReader.cxx b/Detectors/TRD/reconstruction/src/DataReader.cxx
index 5b521b2cfc0..ee913ee1b44 100644
--- a/Detectors/TRD/reconstruction/src/DataReader.cxx
+++ b/Detectors/TRD/reconstruction/src/DataReader.cxx
@@ -65,7 +65,7 @@ WorkflowSpec defineDataProcessing(ConfigContext const& cfgc)
   std::vector<OutputSpec> outputs;
   outputs.emplace_back("TRD", "TRACKLETS", 0, Lifetime::Timeframe);
   outputs.emplace_back("TRD", "DIGITS", 0, Lifetime::Timeframe);
-  outputs.emplace_back("TRD", "TRIGGERRECORD", 0, Lifetime::Timeframe);
+  outputs.emplace_back("TRD", "TRKTRGRD", 0, Lifetime::Timeframe);
   //outputs.emplace_back("TRD", "FLPSTAT", 0, Lifetime::Timeframe);
   LOG(info) << "input spec is:" << inputspec;
   LOG(info) << "enablebyteswap :" << byteswap;
diff --git a/Detectors/TRD/reconstruction/src/DataReaderTask.cxx b/Detectors/TRD/reconstruction/src/DataReaderTask.cxx
index 280f0aeaec0..6e8a58a434b 100644
--- a/Detectors/TRD/reconstruction/src/DataReaderTask.cxx
+++ b/Detectors/TRD/reconstruction/src/DataReaderTask.cxx
@@ -49,7 +49,7 @@ void DataReaderTask::sendData(ProcessingContext& pc)
   LOG(info) << "Sending data onwards with " << mCompressedDigits.size() << " Digits and " << mTracklets.size() << " Tracklets";
   pc.outputs().snapshot(Output{o2::header::gDataOriginTRD, "DIGITS", 0, Lifetime::Timeframe}, mCompressedDigits);
   pc.outputs().snapshot(Output{o2::header::gDataOriginTRD, "TRACKLETS", 0, Lifetime::Timeframe}, mTracklets);
-  pc.outputs().snapshot(Output{o2::header::gDataOriginTRD, "TRIGGERRECORD", 0, Lifetime::Timeframe}, mTriggers);
+  pc.outputs().snapshot(Output{o2::header::gDataOriginTRD, "TRKTRGRD", 0, Lifetime::Timeframe}, mTriggers);
   //    pc.outputs().snapshot(Output{o2::header::gDataOriginTRD,"STATS",0,Lifetime::Timerframe},mStats);
 }
 
```
Before, the datareader was publishing the TriggerRecords with a wrong datatype

bazinski commented:
Hi @davidrohr I had lots of problems with the range based loops, and busy removing them. It will get optimised now that its working properly. The excess memcpy will get fixed. Either in this pr or the next one?

davidrohr commented:
> Hi @davidrohr I had lots of problems with the range based loops, and busy removing them. It will get optimised now that its working properly. The excess memcpy will get fixed. Either in this pr or the next one?

OK, then just apply my fix for know, it should fix it hopefully. Without it is practically unusable.

bazinski committed:
remove range based loops from event record and fix trigger name in message

bazinski commented:
eventrecord fixed and my example 100 events time goes from 1394ms to 74ms.
@davidrohr your patch applied.

davidrohr commented on Detectors/TRD/workflow/CMakeLists.txt:
```diff
@@ -48,12 +48,6 @@ o2_add_executable(trd-vdrift-exb
                   SOURCES src/trd-calib-workflow.cxx
                   PUBLIC_LINK_LIBRARIES O2::Framework O2::TRDCalibration O2::TRDWorkflow O2::DetectorsCalibration)
 
-
```
This should not be removed

bazinski replied:
that was a bad rebase, and by mistake included it then removed it.

bazinski replied:
just clarity the trackreader moved to io

martenole replied:
But this should not appear in this PR, right?

davidrohr reviewed (changes requested):
The tracklets work for me correctly now, but I think the PR was not correctly rebased, why is the track-reader removed from CMake?

davidrohr reviewed (dismissed):
actually, it seems it is added forth and back in 2 commits in this PR. So anyway, in order to not delay it merge, let's merge it as is as soon as the CI passes.

bazinski commented:
trackreader was complaining and not in dev branch

bazinski commented:
by mistake this now includes the digit limitation.
~~trap2raw now has an additional command line parameter of digitrate (default=1000) and limits the raw data generation to only have digits in every 1000th event include event 0 so i can test.~~

~~Currently waiting for simulation for 2000 events to finsh, to check the default is real. The hex dumps look right with tracklets and no digits where i looked, and can be parsed.~~

bazinski commented:
removed the digit limitation to its own branch incase there is a bug found when simulation is finished

davidrohr commented:
Error is unrelated, let's get this merged.
